lnwallet/chanfunding/coin_select.go

```diff
@@ -262,70 +252,85 @@ func CoinSelectSubtractFees(feeRate chainfee.SatPerKWeight, amt,
 	btcutil.Amount, error) {
 
 	// Get balance from coins.
-	var balance btcutil.Amount
+	var totalBalance btcutil.Amount
 	for _, coin := range coins {
-		balance += btcutil.Amount(coin.Value)
+		totalBalance += btcutil.Amount(coin.Value)
 	}
 
-	// First we try to select coins up to the specified maxAmt.
-	outputAmt := maxAmt
+	// First we try to select coins to create an output of the specified
+	// maxAmount and a change output that covers the miner fee.
 	selectedCoins, changeAmt, err := CoinSelect(
-		feeRate, maxAmt, dustLimit, coins,
+		feeRate, maxAmount, dustLimit, coins,
 	)
-	if err == nil {
+
+	outputAmount := maxAmount
+
+	switch err.(type) {
+	// If that selection succeeds we check if our total balance covers
+	// the selected set of coins including fees plus an optional anchor
+	// reserve.
+	case nil:
 		// Validate that the selected amount will leave the required reserved
-		// value in the balance.
+		// value in our total balance.
+		// First we sum up the value of all selected coins.
 		selected := btcutil.Amount(0)
 		for _, coin := range selectedCoins {
 			selected += btcutil.Amount(coin.Value)
 		}
 
-		// We subtract the change amount to obtain the actual amount that is
-		// selected.
+		// We then subtract the change amount from the value of all selected
+		// coins to obtain the actual amount that is selected.
 		selected -= changeAmt
 
-		if balance-selected < reserved {
-			err = &errReservedValueInvalidated{reserved}
-		}
-	}
-	if err != nil {
-		switch err.(type) {
-		// Without sufficient funds we spend it all by subtracting
-		// the fee directly from the output amount.
-		case *ErrInsufficientFunds, *errReservedValueInvalidated:
-			selectedCoins, outputAmt, changeAmt, err = CoinSelectSubtractFees(
-				feeRate, balance-reserved, dustLimit, coins,
-			)
+		// Next we check if our total balance can cover for the selected
+		// output plus the optional anchor reserve.
+		if totalBalance-selected < reserved {
+			// If our local balance is insuffcient to cover for the reserve
+			// we try in the next step to select an output amount that uses
+			// our total balance minus reserve and fees.
+			selectedCoins, outputAmount, changeAmt, err =
+				CoinSelectSubtractFees(
+					feeRate, totalBalance-reserved, dustLimit, coins,
+				)
 			if err != nil {
 				return nil, 0, 0, err
 			}
-
-		default:
+		}
+	// If the initial coin selection fails due to insufficient funds we
+	// select our total available balance minus fees and optional reserve.
+	case *ErrInsufficientFunds:
+		selectedCoins, outputAmount, changeAmt, err = CoinSelectSubtractFees(
+			feeRate, totalBalance-reserved, dustLimit, coins,
+		)
+		if err != nil {
 			return nil, 0, 0, err
 		}
+	default:
+		return nil, 0, 0, err
 	}
 
 	// Sanity check: Always make sure the output amount is below the
 	// specified max amount.
-	switch d := outputAmt - maxAmt; {
-	case d > dustLimit:
+	switch overshoot := outputAmount - maxAmount; {
```

Comment:
Can you squash all changes in this commit that touch code that was introduced in previous commits in the same PR? Makes it easier to review if the same code isn't touched multiple times in the same PR.

Reply:
Done.

Reply:
Done